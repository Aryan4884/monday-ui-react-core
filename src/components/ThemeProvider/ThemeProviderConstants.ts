export enum SystemTheme {
  LIGHT = "light",
  DARK = "dark",
  BLACK = "black",
  HACKER = "hacker"
}

// Colors which are eligible for theming
export enum Color {
  primaryColor = "primary-color",
  primaryHoverColor = "primary-hover-color"
}

export type Theme = {
  name: string;
  colors: SystemThemeColorMap;
};

type SystemThemeColorMap = {
  [key in SystemTheme]?: ColorTokenValueMap;
};

type ColorTokenValueMap = {
  [key in Color]?: string;
};

export const SystemThemeClassMap: SystemThemeClassMapType = {
  [SystemTheme.LIGHT]: "light-app-theme",
  [SystemTheme.DARK]: "dark-app-theme",
  [SystemTheme.BLACK]: "black-app-theme",
  [SystemTheme.HACKER]: "hacker_theme-app-theme"
};

type SystemThemeClassMapType = {
  [key in SystemTheme]: string;
};
