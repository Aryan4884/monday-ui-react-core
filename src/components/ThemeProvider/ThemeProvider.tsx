import React, { FC, ReactElement, useLayoutEffect } from "react";
import { Color, SystemTheme, SystemThemeClassMap, Theme } from "./ThemeProviderConstants";
import cx from "classnames";

export interface ThemeProviderProps {
  theme?: Theme;
  children: ReactElement;
}

const getThemeProviderClassname = (themeName: string) => {
  if (!themeName) {
    return null;
  }
  return `theme-provider_${themeName}`;
};

const generateThemeCssOverride = (theme: Theme) => {
  let css = "";
  for (const systemTheme of Object.keys(theme.colors) as SystemTheme[]) {
    for (const colorToken of Object.keys(theme.colors[systemTheme]) as Color[]) {
      css += `
        .${SystemThemeClassMap[systemTheme]} {
          .${getThemeProviderClassname(theme.name)} {
            --${colorToken}: ${theme.colors[systemTheme][colorToken]};
          }
        }
      `;
    }
  }

  return css;
};

const ThemeProvider: FC<ThemeProviderProps> = ({ theme, children }) => {
  useLayoutEffect(() => {
    if (!theme || !theme.colors) {
      return;
    }

    // Create a new style element
    const styleElement = document.createElement("style");
    styleElement.type = "text/css";
    const themeCssOverride = generateThemeCssOverride(theme);

    try {
      // Set the innerText property of the style element with the dynamic CSS
      styleElement.appendChild(document.createTextNode(themeCssOverride));

      // Append the style element to the document head
      document.head.appendChild(styleElement);
    } catch (error) {
      console.error("Error loading dynamic CSS:", error);
    }

    // Cleanup: Remove the style element when the component unmounts
    return () => {
      document.head.removeChild(styleElement);
    };
  }, [theme]);

  // Pass the theme name as a class to the children - to scope the effect of the theme
  return React.cloneElement(children, {
    className: cx(getThemeProviderClassname(theme?.name), children?.props?.className)
  });
};

export default ThemeProvider;
