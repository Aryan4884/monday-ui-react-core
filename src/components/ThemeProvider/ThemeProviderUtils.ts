import { Theme, ThemeColorTokenValueMap, SystemTheme, SystemThemeClassMap } from "./ThemeProviderConstants";

const generateCss = (object: ThemeColorTokenValueMap, stack: string, parentSelector: string) => {
  for (const key of Object.keys(object)) {
    if (typeof object[key as keyof ThemeColorTokenValueMap] === "string") {
      stack += `--${key}: ${object[key as keyof ThemeColorTokenValueMap]};`;
    }
  }

  if (stack !== "") {
    stack = parentSelector + " {" + stack + "}";
  }

  for (const key of Object.keys(object)) {
    if (typeof object[key as keyof ThemeColorTokenValueMap] === "object") {
      const selector = `${parentSelector} .${key}`;
      stack +=
        "\n" + generateCss(object[key as keyof ThemeColorTokenValueMap] as ThemeColorTokenValueMap, "", selector);
    }
  }

  return stack;
};

export const shouldGenerateTheme = (theme: Theme) => {
  return !!theme?.colors;
};

export const generateThemeCssOverride = (theme: Theme) => {
  if (!shouldGenerateTheme(theme)) {
    return null;
  }

  let css = "";
  for (const systemTheme of Object.keys(theme.colors) as SystemTheme[]) {
    css += generateCss(theme.colors[systemTheme], "", `.${SystemThemeClassMap[systemTheme]} .${theme.name}`) + "\n";
  }

  return css;
};
