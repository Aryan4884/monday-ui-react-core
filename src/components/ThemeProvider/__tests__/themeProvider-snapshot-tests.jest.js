import React from "react";
import renderer from "react-test-renderer";
import ThemeProvider from "../ThemeProvider";

const THEME_NAME = "test-theme-name";

describe("ThemeProvider renders correctly", () => {
  it("with theme name", () => {
    const tree = renderer
      .create(
        <ThemeProvider
          theme={{
            name: THEME_NAME
          }}
        >
          <div />
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
