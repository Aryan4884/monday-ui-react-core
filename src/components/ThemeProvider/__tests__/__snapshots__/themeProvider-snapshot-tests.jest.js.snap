// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider renders correctly with theme name 1`] = `
<div
  className="test-theme-name"
/>
`;
